packages/utils/src/surescripts/send-customer-request.ts

```diff
@@ -0,0 +1,103 @@
+import { Command } from "commander";
+import { out } from "@metriport/core/util/log";
+import { SurescriptsSftpClient } from "@metriport/core/external/surescripts/client";
+import { SurescriptsDataMapper } from "@metriport/core/external/surescripts/data-mapper";
+import { writeSurescriptsRunsFile, appendToSurescriptsRunsFile } from "./shared";
+
+/**
+ * Sends a request for all patients of the given customer, batching requests by facility. This is the preferred method
+ * to accurately send a backfill request to Surescripts for all patients with the minimum number of SFTP transactions.
+ *
+ * Usage:
+ * npm run surescripts -- customer-request --cx-id <cx-id> --csv-output <csv-output> --batch-size <batch-size>
+ *
+ * cx-id: The CX ID of the requester
+ * csv-output: The file to write CSV IDs to
+ * batch-size: The maximum number of patients per batch request - defaults to 100
+ *
+ * Example:
+ * npm run surescripts -- customer-request --cx-id "acme-uuid-1234-sadsjksl" --csv-output "acme-roster.csv" --batch-size 100
+ *
+ * Note: The headers of the CSV output file are "facility_id","transmission_id","patient_id", which are used in scripts
+ * like `convert-customer-responses` to convert the eventual responses from Surescripts into FHIR resources.
+ */
+const program = new Command();
+
+program
+  .name("customer-request")
+  .requiredOption("--cx-id <cx>", "The CX ID of the requester")
+  .requiredOption("--csv-output <csvOutput>", "The file to write CSV IDs to")
+  .option("--batch-size <batchSize>", "The maximum number of patients per batch request", "100")
+  .description(
+    "Send a request to Surescripts for all patients of the given customer, batched by facility"
+  )
+  .showHelpAfterError()
+  .version("1.0.0")
+  .action(
+    async ({
+      cxId,
+      batchSizeString,
+      csvOutput,
+    }: {
+      cxId: string;
+      batchSizeString?: string;
+      csvOutput: string;
+    }) => {
+      const batchSize = parseInt(batchSizeString ?? "100");
+      if (!Number.isFinite(batchSize)) {
+        throw new Error("Batch size must be a number");
+      }
+      const { log } = out(`Surescripts sendCustomerRequest - cxId ${cxId}`);
+      const dataMapper = new SurescriptsDataMapper();
+      const batchRequests = await dataMapper.getBatchRequestDataByFacility(cxId, batchSize);
+      const facilityIds = Object.keys(batchRequests);
+      log(`Sending requests across ${facilityIds.length} facilities`);
+
+      const csvOutputPath = writeSurescriptsRunsFile(
+        csvOutput,
+        `"facility_id","transmission_id","patient_id"\n`
+      );
+      let totalRequestedPatients = 0;
+      let totalBatches = 0;
+
+      for (const facilityId of facilityIds) {
+        const batchesForFacility = batchRequests[facilityId];
+        if (!Array.isArray(batchesForFacility)) {
+          log(`No requests for facility ${facilityId}`);
+          continue;
+        }
+
+        for (const batchRequest of batchesForFacility) {
```

Comment:
This is seems a little circuitous, in`dataMapper.getBatchRequestDataByFacility` we create a somewhat complex data structure just to index into it again. Why do we need to build a map thats organized by facility? Can we create a single array of payloads that we then iterate over and process?

Reply:
I don't understand, if you put the facilityId _in_ the object, then you have a single list of payloads that you iterate over. you'll still have access to the facilityId

Reply:
Ahh I misunderstood the comment earlier, yes good idea and just implemented 👍